### DISCUSSION TITLE
MQTT client disconnected, reason: shutdown,badmatch

### DISCUSSION THREAD

**Discussion by dsilletti** [emqx/emqx#5691] [Q&A]
Hi, my mqtt clients are getting disconnected randomly, sometimes after days, sometimes after few minutes or few hours.
I found this message in the logs:

```{"log":"2021-09-08T14:46:15.538241+00:00` [info] CLIENT_ID@IP_ADDRESS:PORT file: emqx_connection.erl, line: 520, mfa: {emqx_connection,terminate,2}, msg: terminate, pid: \u003c0.3941.0\u003e, reason: `{shutdown,{badmatch,{#{},\u003c\u003c\"14.08,315.76,317.2,317.88,318.07,318.02,317.71,\"\u003e\u003e}}}\n","stream":"stdout","time":"2021-09-08T14:46:15.538549766Z"}```

the broker is EMQX Broker v4.3.6 official docker container
the clients are ESP32 using Arduino core v2.0.0 (also with the 1.x I had the same issue) and the MQTT library is pubsubclient (https://github.com/knolleary/pubsubclient)

Could you please help me to understand the reason behind the disconnection? Thanks

**Comment by dsilletti** [comment]
more info from debug level logs:

```{"log":"2021-09-08T22:45:58.655887+00:00` [error] CLIENT_ID@IP_ADDRESS:PORT [MQTT] , Parse failed for {badmatch,{#{},\u003c\u003c\"-27.42,-13.76,-1.22,10.63,23.4,36.7,49.14,60.97\"\u003e\u003e}}, [{emqx_frame,parse_packet,3,[{file,\"emqx_frame.erl\"},{line,239}]},{emqx_frame,parse_frame,4,[{file,\"emqx_frame.erl\"},{line,181}]},{emqx_connection,parse_incoming,3,[{file,\"emqx_connection.erl\"},{line,625}]},{emqx_connection,handle_msg,2,[{file,\"emqx_connection.erl\"},{line,618}]},{emqx_connection,process_msg,2,[{file,\"emqx_connection.erl\"},{line,364}]},{emqx_connection,handle_recv,3,[{file,\"emqx_connection.erl\"},{line,328}]},{proc_lib,wake_up,3,[{file,\"proc_lib.erl\"},{line,236}]}], Frame data:\u003c\u003c\".11,-27.42,-13.76,-1.22,10.63,23.4,36.7,49.14,60.97,73.14,85.66,98.21,110.01,121.04,132.48,144.07,154.91,165.31,175.71,186.16,196.6,206.75,216.45,226.02,235.9,245.09,252.26,258.7,264.63,267.91,271.67,278.06,283.91,289.72,295.78,300.48,304.87,309.07,312.1,314.08,315.76,317.2,317.88,318.07,318.02,317.71,317.29,316.72,315.89,314.56,312.36,308.57,303.24,299.06,296.42,292.81,288.15,282.19,274.34,26\"...\u003e\u003e\n","stream":"stdout","time":"2021-09-08T22:45:59.036244787Z"}```

```{"log":"2021-09-08T22:45:59.424073+00:00 [debug] CLIENT_ID@IP_ADDRESS:PORT [MQTT] Force to close the socket due to {badmatch,{#{},\u003c\u003c\"-27.42,-13.76,-1.22,10.63,23.4,36.7,49.14,60.97\"\u003e\u003e}}\n","stream":"stdout","time":"2021-09-08T22:45:59.434456626Z"}```

```{"log":"2021-09-08T22:45:59.425565+00:00 [info] CLIENT_ID@IP_ADDRESS:PORT file: emqx_connection.erl, line: 520, mfa: {emqx_connection,terminate,2}, msg: terminate, pid: \u003c0.2026.0\u003e, reason: {shutdown,{badmatch,{#{},\u003c\u003c\"-27.42,-13.76,-1.22,10.63,23.4,36.7,49.14,60.97\"\u003e\u003e}}}\n","stream":"stdout","time":"2021-09-08T22:45:59.437495935Z"}```

```{"log":"2021-09-08T22:45:59.426039+00:00 [error] supervisor: 'esockd_connection_sup - \u003c0.1862.0\u003e', errorContext: connection_shutdown, reason: {badmatch,{#{},\u003c\u003c\"-27.42,-13.76,-1.22,10.63,23.4,36.7,49.14,60.97\"\u003e\u003e}}, offender: [{pid,\u003c0.2026.0\u003e},{name,connection},{mfargs,{emqx_connection,start_link,[[{deflate_options,[]},{max_conn_rate,1000},{active_n,100},{zone,external},{proxy_address_header,\u003c\u003c\u003e\u003e},{proxy_port_header,\u003c\u003c\u003e\u003e},{supported_subprotocols,[]}]]}}]\n","stream":"stdout","time":"2021-09-08T22:45:59.437542764Z"}```

Any suggestion on what is the cause of this disconnection? Thanks

**Comment by zmstone** [comment]
Seems to me the client maybe sent less bytes than they should in the previous packet.
So the remaining bytes are taken by the broker as a new packet, which happened to be parsed as a CONNACK, which doesn’t make any sense because clients should never send it, it’s a server to client message.

These bytes look like payload, but they were parsed as a part of a CONNACK packet.
`\u003c\u003c\"-27.42,-13.76,-1.22,10.63,23.4,36.7,49.14,60.97\"\u003e\u003e`

**Reply by dsilletti** [reply]
thank you @zmstone, I will investigate on the client side, in the while is there a way to ignore the badmatch to avoid the disconnection?

**Reply by zmstone** [reply]
Hi. My description was not quite accurate ,
The client might have sent a wrong packet length followed with more bytes than the length specified.

MQTT messages are parsed out of the tcp stream, when length- body is misaligned, all the following bytes will be garbled (lack of a clear boundary where to start parsing the next packet).

The only way to recover is to close the connection and start over. (which is essentially what the crash triggers)
